planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/ConfiguredFeature.java

```diff
@@ -0,0 +1,302 @@
+  }
+
+  /**
+   * Performs data type normalization in cases where there are value variations. For example, yes/1/true all mean the
+   * same thing if a tag is boolean in nature.
+   * 
+   * @param attributeDataType data type to normalize
+   * @return lambda that converts raw tag values to normalized attribute values
+   */
+  private static Function<Object, Object> typeConverter(AttributeDataType attributeDataType) {
+
+    if (Objects.isNull(attributeDataType)) {
+      return in -> {
+        return in;
+      };
+    }
+
+    switch (attributeDataType) {
+      case bool:
+        return booleanTypeConverter();
+      //Default: pass through
+      default:
+        return in -> in;
+    }
+  }
+
+  private static Set<String> booleanTrue = new HashSet<>(Arrays.<String>asList("true", "yes", "1"));
+  private static Set<String> booleanFalse = new HashSet<>(Arrays.<String>asList("false", "no", "0"));
+
+  /**
+   * Convert a raw tag value to a corresponding normalized boolean value.
+   * 
+   * @return normalization function
+   */
+  private static Function<Object, Object> booleanTypeConverter() {
+    return s -> {
+      if (booleanTrue.contains(s)) {
+        return Boolean.TRUE;
+      }
+      if (booleanFalse.contains(s)) {
+        return Boolean.FALSE;
+      }
+      return null;
+    };
+  }
+
+  /**
+   * Generate logic which determines the minimum zoom level for a feature based on a configured pixel size limit.
+   * 
+   * @param minTilePercent - minimum percentage of a tile that a feature must cover to be shown
+   * @param minZoom        - global minimum zoom for this feature
+   * @return minimum zoom function
+   */
+  private static Function<SourceFeature, Integer> attributeZoomThreshold(double minTilePercent, int minZoom) {
+    return sf -> {
+      try {
+        int zoom = (int) (Math.log(minTilePercent / sf.area()) / LOG4);
+        return Math.max(minZoom, Math.min(zoom, 14));
+      } catch (GeometryException e) {
+        return 14;
+      }
+    };
+  };
+
+  /**
+   * Generates a function which produces a fully-configured attribute for a feature.
+   * 
+   * @param attribute - configuration for this attribute
+   * @return processing logic
+   */
+  private BiConsumer<SourceFeature, Feature> attributeProcessor(AttributeDefinition attribute) {
+    String tagKey = attribute.getKey();
+    Integer configuredMinZoom = attribute.getMinZoom();
+
+    int minZoom = configuredMinZoom == null ? 0 : configuredMinZoom.intValue();
+    Function<SourceFeature, Object> attributeValueProducer = attributeValueProducer(attribute);
+
+    TagCriteria attrIncludeWhen = attribute.getIncludeWhen();
+    TagCriteria attrExcludeWhen = attribute.getExcludeWhen();
+
+    Predicate<SourceFeature> attributeTest =
+      attrIncludeWhen == null ? sf -> true : attrIncludeWhen.matcher();
+    Predicate<SourceFeature> attributeExcludeTest =
+      attrExcludeWhen == null ? sf -> true : attrExcludeWhen.matcher().negate();
```

Comment:
We should use `Expression` instead of `matcher()` as well.  We might even want to use an indexed `MultiExpression` that returns the attribute processors to check based on the element's tags.